### PR TITLE
Introduce SpeziNotifications

### PR DESCRIPTION
**Pull request by bauer-andreas**
# Introduce SpeziNotifications

## :recycle: Current situation & Problem
This PR marks the first release of the SpeziNotifications package. It introduces the `Notifications` Module and introduces two new actions: `notificationSettings` to retrieve the current notification settings (e.g., authorization status) and `requestNotificationAuthorization`. All notification related actions are accessible through the `@Application` property wrapper inside Spezi Modules and the `@Environment` property wrapper inside SwiftUI Views. The `registerRemoteNotifications` and `unregisterRemoteNotifications` actions are now also available via the `@Environment` property wrapper.
Some infrastructure is still exported through Spezi (see https://github.com/StanfordSpezi/Spezi/pull/117).

## :gear: Release Notes 
Each target exposes a structured DocC documentation catalog and documents all public interfaces.

## :books: Documentation
* New `@Application(\.notificationSettings)` and `@Environment(\.notificationSettings)` action to retrieve current notification settings like the authorization status.
* New `@Application(\.requestNotificationAuthorization)` and `@Environment(\. requestNotificationAuthorization)` action to easily request notification authorization.
* New `Notifications` module that makes it easier to interact with notification related actions in Spezi application and ensures to silence concurrency warnings.
* New `XCTSpeziNotifications` target to easily handle notification authorization alerts in UI tests.
* New `XCTSpeziNotificationsUI` target that can be used in UI test application to visualize pending notification requests.

## :white_check_mark: Testing
Added extensive unit and UI testing for all components.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,55 +1,37 @@
...
-[![](https://img.shields.io/endpoint?url=https%3A%2F%2Fswiftpackageindex.com%2Fapi%2Fpackages%2FStanfordBDHG%2FSwiftPackageTemplate%2Fbadge%3Ftype%3Dplatforms)](https://swiftpackageindex.com/StanfordBDHG/SwiftPackageTemplate)
+[![Build and Test](https://github.com/StanfordSpezi/SpeziNotifications/actions/workflows/build-and-test.yml/badge.svg)](https://github.com/StanfordSpezi/SpeziNotifications/actions/workflows/build-and-test.yml)
+[![codecov](https://codecov.io/gh/StanfordSpezi/SpeziNotifications/graph/badge.svg?token=dWaDzUBFoV)](https://codecov.io/gh/StanfordSpezi/SpeziNotifications)
 
+Simplify User Notifications in Spezi-based applications.
 
-## How To Use This Template
+## Overview
+ 
+TODO
```
Comment (bauer-andreas) [inline README.md]: Add Docs overview here and in the DocC catalog.

File: .spi.yml [hunk 2]
```diff
@@ -11,4 +11,6 @@ builder:
   configs:
     - platform: ios
       documentation_targets:
-      - TemplatePackage
+      - SpeziNotifications
+      - XCTSpeziNotifications
```
Comment (PSchmiedmayer) [inline .spi.yml:15]: Very good job with the additional targets!!

File: README.md (hunk 1 above)
Comment (PSchmiedmayer) [reply README.md]: Thank you for working on this!

File: README.md [hunk 3]
```diff
@@ -1,55 +1,37 @@
...
-## Installation
-
-The project can be added to your Xcode project or Swift Package using the [Swift Package Manager](https://github.com/apple/swift-package-manager).
-
-**Xcode:** For an Xcode project, follow the instructions on [adding package dependencies to your app](https://developer.apple.com/documentation/xcode/adding-package-dependencies-to-your-app).
-
-**Swift Package:** You can follow the [Swift Package Manager documentation about defining dependencies](https://github.com/apple/swift-package-manager/blob/main/Documentation/Usage.md#defining-dependencies) to add this project as a dependency to your Swift Package.
+## Setup
 
+You need to add the SpeziDevices Swift package to
```
Comment (PSchmiedmayer) [inline README.md]:
```suggestion
You need to add the SpeziNotifications Swift package to
```

File: README.md [hunk 4]
```diff
@@ -1,55 +1,37 @@
...
 
 
 ## Contributors
-This project is developed as part of the Stanford Byers Center for Biodesign at Stanford University.
-See [CONTRIBUTORS.md](https://github.com/StanfordBDHG/TemplatePackage/tree/main/CONTRIBUTORS.md) for a full list of all TemplatePackage contributors.
+This project is developed as part of the Stanford Mussallem Center for Biodesign at Stanford University.
+See [CONTRIBUTORS.md](https://github.com/StanfordBDHG/StanfordSpezi/tree/main/CONTRIBUTORS.md) for a full list of all SpeziNotifications contributors.
 
 ![Stanford Byers Center for Biodesign Logo](https://raw.githubusercontent.com/StanfordBDHG/.github/main/assets/biodesign-footer-light.png#gh-light-mode-only)
 ![Stanford Byers Center for Biodesign Logo](https://raw.githubusercontent.com/StanfordBDHG/.github/main/assets/biodesign-footer-dark.png#gh-dark-mode-only)
```
Comment (PSchmiedmayer) [inline README.md]: Could also update the alt-text to Mussallem 👍

File: Sources/SpeziNotifications/Actions/NotificationSettings.swift [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    /// ### Action
+    /// - ``NotificationSettingsAction``
+    public var notificationSettings: NotificationSettingsAction {
+        NotificationSettingsAction()
+    }
+}
+
+
+extension EnvironmentValues {
+    /// Retrieve the current notification settings of the application.
```
Comment (PSchmiedmayer) [inline Sources/SpeziNotifications/Actions/NotificationSettings.swift:57]: Thank you for all the documentation!

File: Sources/SpeziNotifications/Notifications.swift [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+/// - ``pendingNotificationsLimit``
+/// - ``remainingNotificationLimit(isolation:)``
+///
+/// ### Fetching Notifications
+/// - ``pendingNotificationRequests(isolation:)``
+/// - ``deliveredNotifications(isolation:)``
+///
+/// ### Categories
+/// - ``add(isolation:categories:)``
+public final class Notifications: Module, DefaultInitializable, EnvironmentAccessible {
```
Comment (PSchmiedmayer) [inline Sources/SpeziNotifications/Notifications.swift:38]: Amazing abstraction layer!!

File: Sources/SpeziNotifications/SpeziNotifications.docc/SpeziNotifications.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+SPDX-License-Identifier: MIT
+
+-->
+
+Simplify User Notifications in Spezi-based applications.
+
+## Overview
+
+TODO
```
Comment (PSchmiedmayer) [inline Sources/SpeziNotifications/SpeziNotifications.docc/SpeziNotifications.md]: Todo to address this section of the documentation

File: Sources/XCTSpeziNotificationsUI/NotificationRequestView.swift [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+//
+
+import SpeziViews
+import SwiftUI
+import UserNotifications
+
+
+/// Present the details of a notification request.
+@available(iOS 18.0, macOS 15.0, tvOS 18.0, watchOS 11.0, visionOS 2.0, *)
+public struct NotificationRequestView: View {
```
Comment (PSchmiedmayer) [inline Sources/XCTSpeziNotificationsUI/NotificationRequestView.swift:16]: Very cool mechanism to enable an easier testing and debugging!

**Review by PSchmiedmayer — APPROVED** [review]
Amazing job with the initial setup; the testing targets and UI are amazing!!

Happy to see this merged and feel free to do the SPI setup 👍
I can do the Zenodo setup once we have tagged a first release 🚀

**Pull request merged**